PowerSimData2Switch.py

```diff
@@ -0,0 +1,432 @@
+    grid.gencost['after']['p2'].sub(grid.gencost['after']['p1']), fill_value = 0.0)
+df_gen_cost_ratio = df_gen_cost_ratio.fillna(value = 0.0)
+
+df_gen_fix_cost = (grid.gencost['after']['f1']).divide(grid.plant['Pmax'], fill_value = 0.0)
+df_gen_fix_cost = df_gen_fix_cost.multiply(8760, fill_value = 0.0)
+df_gen_fix_cost = df_gen_fix_cost.fillna(value = 0.0)
+
+df_gen_fuel_related = df_gen_cost_ratio.multiply(0.7)
+df_gen_vom = df_gen_cost_ratio.multiply(0.3)
+
+bus_type_fuelcost['AveFuelRelated'] = df_gen_fuel_related
+bus_type_fuelcost['BusAveFuelCost'] = ""
+
+for i in range(len(bus_type_fuelcost.index)):
+    temp_bus_id = bus_type_fuelcost.loc[bus_type_fuelcost.index[i]].at['bus_id']
+    temp_fuel = bus_type_fuelcost.loc[bus_type_fuelcost.index[i]].at['fuel']
+    bus_type_fuelcost.at[bus_type_fuelcost.index[i], 
+                         'BusAveFuelCost'] = df_ave_fuel_cost.loc[temp_bus_id].at[temp_fuel]
+    
+bus_type_fuelcost['Calc_HeatRate'] = bus_type_fuelcost['AveFuelRelated'].divide(
+    bus_type_fuelcost['BusAveFuelCost'], fill_value = 0.0)
+bus_type_fuelcost = bus_type_fuelcost.fillna(value = 0.0)
+
+
+def AveFuelCost(bus_id, fuel, year):
+    
+    """Calculate average fuel cost by type on each bus.
+       :param int64 bus_id: bus number ID.
+       :param str fuel: fuel type.
+       :param str year: investment year.
+       :return: (*float*) -- average fuel cost for a certain fuel type on each bus.
+    """
+    
+    ave_fuel_cost = df_ave_fuel_cost.at[bus_id, fuel]
+    ave_fuel_cost = ave_fuel_cost * (interest_rate**(int(year) - int(base_year)))
+    ave_fuel_cost = round(ave_fuel_cost,2)
+    return ave_fuel_cost
+
+def CalcHeatRate(plant_id):
+    
+    """Calculate heat rate for a certain plant.
+       :param int64 plant_id: plant ID.
+       :return: (*float*) -- heat rate for a plant.
+    """    
+    
+    calc_heat_rate = bus_type_fuelcost.at[plant_id, 'Calc_HeatRate']
+    calc_heat_rate = round(calc_heat_rate,2)
+    if plant_id in ret_unit_list:
+        calc_heat_rate = 999
+    if calc_heat_rate == 0.0:
+        calc_heat_rate = '.'
+    return calc_heat_rate
+
+def GIS_distance(branch_id):
+    
+    """Calculate length for a given branch ID.
+       :param int64 branch_id: branch ID.
+       :return: (*float*) -- straight line distance for a branch.
+    """
+    
+    pi = float (3.14159265354)
+    r = float (3958.756)
+    e = float (1.60934)
+    lat1 = grid.branch.at[branch_id, 'from_lat']
+    long1 = grid.branch.at[branch_id, 'from_lon']
+    lat2 = grid.branch.at[branch_id, 'to_lat']
+    long2 = grid.branch.at[branch_id, 'to_lon']
+    if lat1 == lat2 and long1 == long2:
+        distance = 0.001
+    else:
+        distance = m.acos (m.sin (lat1 / 180 * pi) * m.sin (lat2 / 180 * pi) + 
+                           m.cos (lat1 / 180 * pi) * m.cos (lat2 / 180 * pi) * 
+                           m.cos (long1 / 180 * pi - long2 / 180 * pi) ) * r * e
+    distance = round(distance, 2)

```

Comment:
@YifanLi86 I guess I don't understand the logic here: we replace 0 distance by 0.001 for some reason but round it to 0 afterwards?

Reply:
So what should we do? If SWITCH is happy with 0s, I prefer to putting 0s as they were.

Reply:
We can use the `haversine` package to calculate great-circle distances automatically, or import our implementation from PowerSimData if we want to add/use that dependency.

Reply:
@danielolsen are we pair coding now? You read my mind.

Reply:
It seems that for l. 210, it doesn't matter whether we have `distance = 0.001`, `distance = 0`, `distance = -0.001`, or `distance = 0.004`. After l.215, it will be `0`.

Reply:
@danielolsen Yes, I just want to confirm: 0s are okay and won't break anything since from the code it seems we tried to remove 0s but eventually didn't.

Reply:
I think it is clear now. I will proceed with 0s.

Reply:
@BainanXia We can leave all the 0 as it is for now.